text/2972-constrained-naked.md

```diff
@@ -0,0 +1,131 @@
+
+Because naked functions depend upon the calling convention so heavily, inlining of these functions would make code generation extremely difficult. Therefore, we disallow inlining.
+
+Since the `const` and `sym` operands modify neither the stack nor the registers, their use is permitted.
+
+## Examples
+
+This function adds three to a number and returns the result:
+
+```rust
+const THREE: usize = 3;
+
+#[naked]
+pub extern "sysv64" fn add_n(number: usize) -> usize {
+    unsafe {
+        asm!(
+            "add rdi, {}"
+            "mov rax, rdi"
+            "ret",
+            const THREE,
+            options(noreturn)
+        );
+    }
+}
+```
+
+The calling convention is defined as `extern "sysv64"`, therefore we know that the input is in the `rdi` register and the return value is in the `rax` register. The `asm!()` statement contains `options(noreturn)` and therefore we handle the return directly through the `ret` instruction. We can provide a `const` operand since it modifies neither registers nor stack. Since we have strong guarantees about the state of the registers, we can mark this function as safe and wrap the `asm!()` statement in an `unsafe` block.
+
+# Drawbacks
+
+Implementing this will break compatiliby of existing uses of the nightly `#[naked]` attribute. All of these uses likely depend on undefined behavior. If this is a problem, we could simply use a different attribute.
+
+This definition may be overly strict. There is certainly some code that would work without this. The counter argument is that this code relies on undefined behavior and is probably not worth preserving. It might also be possible to reasonably ease the constraints over time.
+
+This proposal requires the use of assembly where it is theoretically possible to use Rust in a naked function. However, practically the use of Rust in naked functions is nearly impossible and relies on extensively undefined behavior.
+
+Adopting this definition changes the invariants of `asm!()`. Currently, all registers not supplied as operands to `asm!()` contain undefined values. This proposal changes this to define that the initial register state is unmodified from the function call. This is an even stronger commitment than merely guaranteeing calling convention conformance, which some may dislike. However, the change to permit defined initial register state applies **only** to the use of `asm!()` as the body of a naked function.
```

Comment:
Is it worth making this difference behavior explicit in `asm!` rather than implicit? Something like `options(naked)`?

Reply:
@bstrie I'm willing to defer to @Amanieu on this one. However, I believe this phenomenon emerges already from the commitment (already enforced by a lint) that a naked function must contain only a single `asm!` block. There was an earlier code generation bug that broke this assumption. But I fixed it in a previous PR.

The important point here is that the compiler behavior is well defined and any failure of that behavior is a bug.

Reply:
I'm not confident that we can invent something better than we have now -- particularly as features are added to `asm!`.

Reply:
Would it make sense for naked functions to use `global_asm!` instead of `asm!`?

Reply:
@bstrie No. Naked functions additionally provide mangling, debuginfo and docs.